src/main/scala/io/iohk/ethereum/nodebuilder/NodeBuilder.scala

```diff
@@ -676,6 +681,27 @@ trait SyncControllerBuilder {
 
 }
 
+trait PortForwardingBuilder {
+  self: DiscoveryConfigBuilder =>
+
+  import monix.execution.Scheduler.Implicits.global
+
+  lazy val startPortForwarding: CancelableFuture[(Unit, Task[Unit])] =
+    PortForwarder
+      .openPorts(
+        Seq(Config.Network.Server.port),
+        Seq(discoveryConfig.port).filter(_ => discoveryConfig.discoveryEnabled)
+      )
+      .whenA(Config.Network.automaticPortForwarding)
+      .allocated
+      .runToFuture
+
+  def stopPortForwarding(): Future[Unit] =
+    startPortForwarding.flatMap { case (_, release) =>
+      release.runToFuture
+    }
+}
```

Comment:
It would be cool if starting and stopping could be done repeatedly, at least in theory. The lazy val can only be used once, but you could use an `AtomicReference` to set/unset it when its started/stopped.

This is what I ended up with in a similar situation where I had to use mutable state with resource allocation:

```scala
  // Reference to an allocated miner and its release task, if it has been started.
  // The release cannot be ignored or it would keep the other resources from being released.
  private val maybeMinerPromiseRef =
    new AtomicReference[Option[Promise[(ActorRef, Task[Unit])]]](None)

  private def startMiningProcess(node: Node): Unit = {
    val minerPromise = Promise[(ActorRef, Task[Unit])]()
    if (maybeMinerPromiseRef.compareAndSet(None, Some(minerPromise))) {

      val miner: Resource[Task, ActorRef] = ???

      miner.allocated.runToFuture
        .flatMap {
          case (actor, release) =>
            minerPromise.success(actor -> release)
            sendMiner(MinerProtocol.StartMining)
        }
        .recover {
          case NonFatal(ex) =>
            minerPromise.tryFailure(ex)
        }
    }
  }

  private def stopMiningProcess(): Unit = {
    sendMiner(MinerProtocol.StopMining).flatMap { _ =>
      maybeMinerPromiseRef
        .get()
        .map { minerPromise =>
          minerPromise.future.flatMap {
            case (_, release) =>
              maybeMinerPromiseRef.set(None)
              release.runToFuture
          }
        }
        .getOrElse(Future.successful(()))
    }
  }
```

Reply:
Seems easy enough 👍

Reply:
Actually I should probably have made the stopping safer as well like this:
```scala
  private def stopMiningProcess(): Unit = {
      maybeMinerPromiseRef
        .getAndSet(None) match {
          case None => Future.unit
          case Some((_, release)) => release.runToFuture
        }
    }
  }
```
So it definitely only executes once.

Reply:
I think there's a potential race condition in your original snippet: if `stopMiningProcess` is called before the promise is updated with the `release` task then this will never run. I think there's not an easy way to prevent this without locking (with eg. `synchronize`), what do you think?

Reply:
Yes once you have the promise instance, you can call `.future` to subscribe to its completion. Just have to make sure it will happen even if there was an error.